### ISSUE TITLE
A managed exception in a callback from native code seg faults on OS-X

### ISSUE THREAD

**Issue by toptensoftware** [dotnet/coreclr#953]
As per title, see sample below or [this project](https://www.dropbox.com/s/s88s9wjkae233z1/clrtest2.zip?dl=0).

```
		[DllImport("clrtest.dylib")]
		public static extern void InvokeCallback(Action callback);

		static void Main(string[] args)
		{
			InvokeCallback(() => {
				throw new NotImplementedException();
			});
		}
```

**Comment by masonwheeler** [comment]
Is this actually a problem?  [Having it crash on you looks like expected behavior to me...](http://blogs.msdn.com/b/junfeng/archive/2008/01/28/reverse-p-invoke-and-exception.aspx)

**Comment by toptensoftware** [comment]
I understand that point, but disagree in this particular case - seg fault shouldn't be the expected behaviour.  If it was to log the exception and kill the process I could live with that, but a seg fault with no info is less than ideal and makes it extremely difficult to diagnose the cause of the exception.

**Comment by janvorli** [comment]
@toptensoftware Seg fault in this case is a bug. It should exit the process and report the exception instead.
It would be great if you could add here a sample code of the InvokeCallback so that this issue is self contained.
I will investigate it.

**Comment by kangaroo** [comment]
@janvorli follow the link to 'this project' in his original post, he's got a zip with the source.

**Comment by kangaroo** [comment]
A little bit additional info before the airline likes me turn off data:

```
(lldb) bt
* thread #1: tid = 0x5d3b, 0x0000000000000004, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x4)
  * frame #0: 0x0000000000000004
    frame #1: 0x00007fff95f58951 libunwind.dylib`_Unwind_RaiseException + 153
    frame #2: 0x00007fff8c305d32 libc++abi.dylib`__cxa_rethrow + 77
    frame #3: 0x0000000101844883 libcoreclr.dylib`StartUnwindingNativeFrames + 64 at unixasmhelpers.S:933
    frame #4: 0x0000000101844a22 libcoreclr.dylib`UMThunkStub + 278 at umthunkstub.S:870
    frame #5: 0x00000001007f1f8f clrtest.dylib`InvokeCallback + 15
    frame #6: 0x00000001027be5a1
    frame #7: 0x00000001027be4f9
    frame #8: 0x000000010184409d libcoreclr.dylib`FastCallFinalizeWorker_End + 124 at calldescrworkeramd64.S:840
    frame #9: 0x0000000101641b63 libcoreclr.dylib`CallDescrWorkerWithHandler(pCallDescrData=0x00007fff5fbfe2f8, fCriticalCall=0) + 467 at callhelpers.cpp:88
    frame #10: 0x0000000101642ddc libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(this=0x00007fff5fbfe540, pArguments=0x00007fff5fbfe4c0) + 2412 at callhelpers.cpp:627
    frame #11: 0x00000001015c513d libcoreclr.dylib`MethodDescCallSite::Call(this=0x00007fff5fbfe540, pArguments=0x00007fff5fbfe4c0) + 29 at callhelpers.h:421
    frame #12: 0x0000000101265eb4 libcoreclr.dylib`RunMain(this=0x00007fff5fbfe658, pParam=0x00007fff5fbfe6f0)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const::'lambda'(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*)::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*) const::Param*) const + 676 at assembly.cpp:2540
    frame #13: 0x0000000101262dfc libcoreclr.dylib`RunMain(this=0x00007fff5fbfe6e0, __EXparam=0x00007fff5fbfe6f0)::$_0::operator()(RunMain(MethodDesc*, short, int*, REF<PtrArray>*)::Param*) const + 76 at assembly.cpp:2558
    frame #14: 0x0000000101262c82 libcoreclr.dylib`RunMain(pFD=0x00000001000e4da0, numSkipArgs=1, piRetVal=0x00007fff5fbfe810, stringArgs=0x00007fff5fbfeb70) + 370 at assembly.cpp:2558
    frame #15: 0x0000000101262fc8 libcoreclr.dylib`Assembly::ExecuteMainMethod(this=0x000000010011b8e8, stringArgs=0x00007fff5fbfeb70) + 328 at assembly.cpp:2666
    frame #16: 0x00000001012f207a libcoreclr.dylib`CorHost2::ExecuteAssembly(this=0x0000000100103b88, dwAppDomainId=1, pwzAssemblyPath=0x000000010011a2a8, argc=0, argv=0x0000000000000000, pReturnValue=0x00007fff5fbff1fc) + 1098 at corhost.cpp:1294
    frame #17: 0x00000001011313c5 libcoreclr.dylib`ExecuteAssembly(exePath=0x0000000100104a40, coreClrPath=0x0000000100104bb0, appDomainFriendlyName=0x0000000100007c43, propertyCount=5, propertyKeys=0x00007fff5fbff680, propertyValues=0x00007fff5fbff650, argc=0, argv=0x0000000000000000, managedAssemblyPath=0x0000000100104b40, entryPointAssemblyName=0x0000000000000000, entryPointTypeName=0x0000000000000000, entryPointMethodName=0x0000000000000000, exitCode=0x00007fff5fbff1fc) + 1829 at unixinterface.cpp:196
    frame #18: 0x000000010000340c corerun`ExecuteManagedAssembly(currentExeAbsolutePath=0x0000000100104a40, clrFilesAbsolutePath=0x0000000100104ae0, managedAssemblyAbsolutePath=0x0000000100104b40, managedAssemblyArgc=0, managedAssemblyArgv=0x0000000000000000) + 3452 at coreruncommon.cpp:275
    frame #19: 0x000000010000753c corerun`corerun(argc=2, argv=0x00007fff5fbffb80) + 2284 at corerun.cpp:150
    frame #20: 0x0000000100007602 corerun`main(argc=2, argv=0x00007fff5fbffb80) + 34 at corerun.cpp:162
    frame #21: 0x00007fff9400a5c9 libdyld.dylib`start + 1
```

**Comment by janvorli** [comment]
I've looked into the issue. The problem is that the UMThunkStub has a personality routine that processes unhandled exceptions in the native code and for some reason, the unwinder fails to get its proper address and gets 4 instead. Trying to call that obviously leads to the crash that can be seen above in the stack trace from @kangaroo.
I am investigating what's wrong.

**Comment by janvorli** [comment]
The issue is caused by the the assembler or linker on OSX ignoring the personality routine address when the personality routine address encoding is 0 (absolute pointer) and always stores 4 as the routine address.
Fortunatelly, using encoding 0x9b (indirect PC relative) works ok on OSX.

**Comment by janvorli** [comment]
Fixed by #986
